### DISCUSSION TITLE
Make `npx directus bootstrap` admin user initialization optional

### DISCUSSION THREAD

**Discussion by Kematia** [directus/directus#6528] [Feature Requests]
Hey! I'm running into a slight problem with the `npx directus bootstrap command`. 

I'm using custom migrations, where I'm spinning up my own roles and admin users. After the bootstrap is run, the CLI attempts to create a default Admin Role and a default Admin user using `ADMIN_NAME` and `ADMIN_PASSWORD` in my .env file, which are the same credentials I use in my migration, leading to this error: 

<img width="1277" alt="Screen Shot 2021-06-28 at 11 04 18" src="https://user-images.githubusercontent.com/17851386/123610153-9f59e380-d800-11eb-8e25-98ffbc34c419.png">

I can omit/rename the `ADMIN_EMAIL` and `ADMIN_PASSWORD` variables, but this leads to a duplicate user and admin role being created:

<img width="917" alt="Screen Shot 2021-06-28 at 11 05 31" src="https://user-images.githubusercontent.com/17851386/123610445-e34ce880-d800-11eb-8a29-f691fa2ff15f.png">

![Screen Shot 2021-06-28 at 11 05 50](https://user-images.githubusercontent.com/17851386/123610485-eb0c8d00-d800-11eb-9ba4-4cc37d8d79e0.png)

I propose adding a check to see if an admin role and/or an admin user exists already, or a flag like `--skipAdminCreation` to the `npx directus bootstrap` command.

Willing to work on this!

edit: while we are at it, another useful flag could be `--skipCustomMigrations` -  I'm using the SDK instead of Knex in some migrations, and the API is not actually running until the migrations are finished. Bit of a niche use case I understand.

**Comment by rijkvanzanten** [comment]
Just to be clear, this only happens when you have a custom migration, that injects a starting user "manually" during the migration?

I think a dedicated flag makes most sense, that makes sure that power users have a way to skip certain steps, while ensuring that regular users don't experience any unexpected side effects based on internal magic 👍🏻 

> I'm using the SDK instead of Knex in some migrations, and the API is not actually running until the migrations are finished. Bit of a niche use case I understand.

I think I'll have to file this under not-supported 🤔 Like you said, the migrations are run "outside" of the API while the API itself isn't running, so you won't be able to access the API during migrations.

**Reply by Kematia** [reply]
> Just to be clear, this only happens when you have a custom migration, that injects a starting user "manually" during the migration?

Affirmative! I load 3 admin accounts, and an editor role. It's ran on every project we manage.

> I think a dedicated flag makes most sense, that makes sure that power users have a way to skip certain steps, while ensuring that regular users don't experience any unexpected side effects based on internal magic 👍🏻

My thinking as well! Will go with that.

> > I'm using the SDK instead of Knex in some migrations, and the API is not actually running until the migrations are finished. Bit of a niche use case I understand.
> 
> I think I'll have to file this under not-supported 🤔 Like you said, the migrations are run "outside" of the API while the API itself isn't running, so you won't be able to access the API during migrations.

I understand your point! 🙂

**Comment by karel-advantitge** [comment]
This is what I was looking for as well.
I see the flag has been added and this has been merged. It would still be quite a hassle to apply this with the docker image though.
How could we make this easier?

The bootstrap that happens in the docker image is this one, if I'm not mistaken: https://github.com/directus/directus/blob/ea19390f11e688911360d1e516a0dbd32db26e44/.github/actions/build-images/rootfs/directus/images/main/rootfs/usr/local/bin/entrypoint#L48

We could conditionally add the flag through an env variable? Though you could argue that we're kind of mixing methodologies if we use bath flags and env variables at once.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
